src/audio/copier/copier_dai.c

```diff
@@ -64,6 +64,92 @@ static int copier_set_alh_multi_gtw_channel_map(struct comp_dev *dev,
 	return 0;
 }
 
+static int copier_alh_assign_dai_index(struct comp_dev *dev,
+				       void *gtw_cfg_data,
+				       union ipc4_connector_node_id node_id,
+				       struct ipc_config_dai *dai,
+				       int *dai_index,
+				       int *dai_count)
+{
+	struct processing_module *mod = comp_get_drvdata(dev);
+	struct copier_data *cd = module_get_private_data(mod);
+	const struct sof_alh_configuration_blob *alh_blob = gtw_cfg_data;
+	uint8_t *dma_config;
+	size_t alh_cfg_size, dma_config_length;
+	int i, dai_num, ret;
+
+	if (!cd->config.gtw_cfg.config_length) {
+		comp_err(mod->dev, "No gateway config found in blob!");
+		return -EINVAL;
+	}
+
+	switch (dai->type) {
+	case SOF_DAI_INTEL_HDA:
+		/* We use DAI_INTEL_HDA for ACE 2.0 platforms */
+		alh_cfg_size = get_alh_config_size(alh_blob);
+		dma_config = (uint8_t *)gtw_cfg_data + alh_cfg_size;
+		dma_config_length = (cd->config.gtw_cfg.config_length << 2) - alh_cfg_size;
+
+		/* Here we check node_id if we need to use FW aggregation,
+		 * in other words do we need to create multiple dai or not
+		 */
+		if (!is_multi_gateway(node_id)) {
+			/* Find DMA config in blob and retrieve stream_id */
+			ret = ipc4_find_dma_config_multiple(dai, dma_config, dma_config_length,
+							    alh_blob->alh_cfg.mapping[0].alh_id, 0);
+			if (ret != 0) {
+				comp_err(mod->dev, "No sndw dma_config found in blob!");
+				return -EINVAL;
+			}
+			dai_index[0] = dai->host_dma_config[0]->stream_id;
+			return 0;
+		}
+
+		dai_num = alh_blob->alh_cfg.count;
+		if (dai_num > IPC4_ALH_MAX_NUMBER_OF_GTW || dai_num < 0) {
+			comp_err(mod->dev, "Invalid dai_count: %d", dai_num);
+			return -EINVAL;
+		}
+
+		for (i = 0; i < dai_num; i++) {
+			ret = ipc4_find_dma_config_multiple(dai, dma_config,
+							    dma_config_length,
+							    alh_blob->alh_cfg.mapping[i].alh_id, i);
+			if (ret != 0) {
+				comp_err(mod->dev, "No sndw dma_config found in blob!");
+				return -EINVAL;
+			}
+			dai_index[i] = dai->host_dma_config[i]->stream_id;
```

Comment:
Please explain the connection between dai_index and stream_id.

Stream_id looks related to the stream_tag used for HDaudio, but it's absolutely not the DAI for SoundWire - which is rather defined in terms of links/PDI.

Or at the very least a comment to say that you intentionally overloaded the concept of DAI but this dai_index is only used for the DMA handling.

Reply:
@plbossart as you mentioned above `stream_tag` used for HDaudio, and because we use HD-A DMA to process data for SoundWire it looked logical for me to use `stream_tag` as a `dai_index` instead of link/PDI.